components/heap/heap_trace_standalone.c

```diff
@@ -304,46 +341,228 @@ static IRAM_ATTR void record_free(void *p, void **callers)
+
+    if (rPrev) {
+        assert(rPrev->next == rRemove);
+    }
+
+    if (rNext) {
+        assert(rNext->prev == rRemove);
+    }
+
+    // update prev neighbor
+    if (rPrev) {
+        rPrev->next = rNext;
+    }
+
+    // update next neighbor
+    if (rNext) {
+        rNext->prev = rPrev;
+    }
+
+    // update first
+    if (rRemove == rs->first) {
+        rs->first = rNext;
+    }
+
+    // update last
+    if (rRemove == rs->last) {
+        rs->last = rPrev;
+    }
+
+    // move self to unused
+    rRemove->next = rs->unused;
+    rRemove->prev = NULL;
+    rs->unused = rRemove;
+
+    // decrement
+    rs->count--;
+
+    if (rs->count > 1) {
+        assert(rs->last != NULL);
+        assert(rs->first != NULL);
+        assert(rs->last != rs->first);
+    }
+
+    if (rs->count == 1) {
+        assert(rs->last != NULL);
+        assert(rs->first != NULL);
+        assert(rs->last == rs->first);
+    }
+    
+    if (rs->count == 0) {
+        assert(rs->last == NULL);
+        assert(rs->first == NULL);
+    }
+}
+
+
+// pop record from unused list
+static IRAM_ATTR heap_trace_record_t* linked_list_pop_unused(const records_t* rs)
+{
+    if (rs->count >= rs->capacity) {
+        return NULL;
+    }
+
+    // we checked that there is capacity, 
+    // so there should be some unused records
+    assert(rs->unused != NULL);
+
+    heap_trace_record_t* pop = rs->unused;
+    assert(pop->address == NULL);
+    assert(pop->size == NULL);
+
+    // update next unused record
+    heap_trace_record_t* next = pop->next;
+
+    // popped last item, next should be NULL
+    if (rs->count == rs->capacity - 1) {
+        assert(next == NULL);
+    }
+
+    // next should be non-null
+    if (rs->count < rs->capacity - 1) {
+        assert(next != NULL);
+        next->prev = NULL;
+    }
+
+    // update unused list
+    rs->unused = next;
+
+    return pop;
+}
+
+// copy a record.
+// Note: only copies the *allocation data*, not the next & prev ptrs
+static IRAM_ATTR void linked_list_copy(heap_trace_record_t *rDest, const heap_trace_record_t *rSrc)
+{
+    rDest->ccount  = rSrc->ccount;
+    rDest->address = rSrc->address;
+    rDest->size    = rSrc->size;
+    memcpy(rDest->freed_by,   rSrc->freed_by,   sizeof(void *) * STACK_DEPTH);
+    memcpy(rDest->alloced_by, rSrc->alloced_by, sizeof(void *) * STACK_DEPTH);
+}
+
+// Append a record to the end of the linked list.
+// This deep copies rAppend into the linked list. 
+static IRAM_ATTR bool linked_list_add(records_t* rs, const heap_trace_record_t *rAppend)
+{
+    if (rs->count < rs->capacity) {
+
+        // get unused record
+        heap_trace_record_t* rDest = linked_list_pop_unused(rs);
+
+        // we checked that there is capacity, so this
+        // should never be null.
+        assert(rDest != NULL);
+        
+        // copy allocation data
+        linked_list_copy(rDest, rAppend);
+
+        // update linked list connectivity
+        rDest->next = NULL;
```

Comment:
It would be better to detach rDest from the unused list in `linked_list_pop_unused` so the function can return an element no longer attached to anything.